src/Saml2/Provider.php

```diff
@@ -0,0 +1,389 @@
+    {
+        if (!empty($key) && empty($this->config[$key])) {
+            return $default;
+        }
+
+        return $key ? Arr::get($this->config, $key, $default) : $this->config;
+    }
+
+    public function redirect()
+    {
+        $this->request->session()->put('state', $state = $this->getState());
+
+        $identityProviderConsumerService = $this->getIdentityProviderEntityDescriptor()
+            ->getFirstIdpSsoDescriptor()
+            ->getFirstSingleSignOnService(SamlConstants::BINDING_SAML2_HTTP_REDIRECT);
+
+        $authnRequest = new AuthnRequest();
+        $authnRequest
+            ->setID(Helper::generateID())
+            ->setProtocolBinding($this->getAssertionConsumerServiceBinding())
+            ->setRelayState($state)
+            ->setIssueInstant(new DateTime())
+            ->setDestination($identityProviderConsumerService->getLocation())
+            ->setNameIDPolicy((new NameIDPolicy())->setFormat(SamlConstants::NAME_ID_FORMAT_PERSISTENT))
+            ->setIssuer(new Issuer($this->getServiceProviderEntityDescriptor()->getEntityID()));
+
+        $redirectBinding = new HttpRedirectBinding();
+
+        $messageContext = new MessageContext();
+        $messageContext->setMessage($authnRequest);
+
+        return $redirectBinding->send($messageContext);
+    }
+
+    protected function getIdentityProviderEntityDescriptorManually(): EntityDescriptor
+    {
+        $acs = $this->getConfig('acs');
+        $entityId = $this->getConfig('entityid');
+        $certificate = $this->getConfig('certificate');
+
+        if (!$entityId || !$certificate) {
+            throw new MissingConfigException('When using "acs", both "entityid" and "certificate" must be set');
+        }
+
+        if (!Str::startsWith($certificate, '-----BEGIN CERTIFICATE-----')) {
+            $certificate = "-----BEGIN CERTIFICATE-----\r\n".
+                implode("\r\n", str_split($certificate, 80)).
+                "\r\n-----END CERTIFICATE-----\r\n";
+        }
+
+        $x509 = new X509Certificate();
+        $x509->loadPem($certificate);
+
+        $builder = new SimpleEntityDescriptorBuilder($entityId, $acs, $acs, $x509);
+
+        return $builder->get();
+    }
+
+    protected function getIdentityProviderEntityDescriptorFromXml(): EntityDescriptor
+    {
+        return Metadata::fromXML($this->getConfig('metadata'), new DeserializationContext());
+    }
+
+    protected function getIdentityProviderEntityDescriptorFromUrl(): EntityDescriptor
+    {
+        $metadataUrl = $this->getConfig('metadata');
+        $xml = Cache::get(self::CACHE_KEY);
+        $ttl = Cache::get(self::CACHE_KEY_TTL);
+
+        $deserializationContext = new DeserializationContext();
+
+        if ($xml && $ttl && $ttl + $this->getConfig('ttl', 86400) > time()) {
+            return Metadata::fromXML($xml, $deserializationContext);
+        }
+
+        Cache::forever(self::CACHE_KEY_TTL, time());
```

Comment:
Couldn't we use an expiring key here, rather than the forever cache?

Reply:
The reason for this is that if your metadata cache is cleared, and due to some network or other issue you cannot refresh the metadata right now then you cannot serve any more authentication requests until you get the metadata again.

There may be a more Laravel idiomatic way of saying "refresh this cached item every X amount of time, but if you fail to refresh it then don't remove it" that I'm not aware of?